src/protocols/MeterModbus.cpp

```diff
@@ -0,0 +1,153 @@
+/**
+ * @file MeterModbus.cpp
+ * @brief
+ *  Created on: Apr 25, 2016
+ * @author Florian Achleitner <florian.achleitner@student.tugraz.at>
+ */
+#include "protocols/MeterModbus.hpp"
+
+MeterModbus::MeterModbus(const std::list<Option> &options) :
+Protocol("modbus"), _libmodbus_debug(false)
+{
+	std::string mb_type = lookup_mandatory<const char *, OptionList::lookup_string>(options, "type", name());
+
+	if (mb_type == "rtu")
+		create_rtu(options);
+	else if (mb_type == "tcp")
+		create_tcp(options);
+	else
+		throw(std::invalid_argument("Modbus type can be rtu or tcp."));
+
+	_libmodbus_debug = lookup_optional<bool, OptionList::lookup_bool>(options, "libmodbus_debug", false);
+
+	std::string map = lookup_mandatory<const char *, OptionList::lookup_string>(options, "register_map", name());
+
+	try {
+	_regmap = RegisterMap::findMap(map);
+	} catch (std::out_of_range &e) {
+		throw(std::invalid_argument("Modbus register map invalid."));
+	}
+}
+
+void MeterModbus::create_rtu(const std::list<Option> &options) {
+	std::string rtu_device;
+	int rtu_baudrate;
+	int rtu_slave;
+
+	rtu_device = lookup_mandatory<const char *, OptionList::lookup_string>(options, "device", name());
+	rtu_baudrate = lookup_optional<int, OptionList::lookup_int>(options, "baudrate", 9600);
+	rtu_slave = lookup_optional<int, OptionList::lookup_int>(options, "slave", 1);
+
+	_mbconn.reset(new ModbusRTUConnection(rtu_device, rtu_baudrate, rtu_slave));
+}
+
+void MeterModbus::create_tcp(const std::list<Option> &options) {
+	std::string tcp_host;
+	int tcp_port;
+	tcp_host = lookup_mandatory<const char *, OptionList::lookup_string>(options, "host", name());
+	tcp_port = lookup_optional<int, OptionList::lookup_int>(options, "port", 502);
+
+	_mbconn.reset(new ModbusTCPConnection(tcp_host, tcp_port));
+}
+
+MeterModbus::~MeterModbus()
+{
+	// TODO Auto-generated destructor stub
+}
+
+int MeterModbus::open() {
+	_mbconn->connect();
+	if (_libmodbus_debug)
+		modbus_set_debug(_mbconn->getctx(), TRUE);
+	return SUCCESS;
+}
+
+int MeterModbus::close() {
+	_mbconn.reset();
+	return SUCCESS;
+}
+
+ssize_t MeterModbus::read(std::vector<Reading> &rds, size_t n) {
+	try {
+		rds = _regmap->read(_mbconn);
```

Comment:
The background is, that I wondered why we pre-create empty Readings in the reading_thread and replace them in the Protocol::read. I assume it's left from the original C codebase. I think this makes no sense, so I am elaborating more C++ish read functions. Those return a vector with all their readings. The size is flexible, and everything would be less verbose. It would look like 
```C++
std::vector<Reading> Meter::read();
```

The Readings are returned by value and moved (C++11 move) on return.

Just like the RegisterMap::read() method. However this is just an idea, and thus MeterModbus::read adapts it to the exisiting scheme in a hacky way ;)

A Reading can be created like
```C++
rds.push_back(Reading(value, new StringIdentifier("TotalExpWh")))
```
instead of 
```C++
if (ret < n) {
	rds[ret].identifier(new ObisIdentifier("2.7.0"));
	rds[ret].value(get_record_value(record));
	rds[ret].time();
	++ret;
 }
```